### ISSUE TITLE
hyphen and dash aren't interchangeable in dask config keys

### ISSUE THREAD

**Issue by ntabris** [dask/dask#9129]
The docs say

> Note that the get function treats underscores and hyphens identically

(https://docs.dask.org/en/stable/configuration.html)

Strictly speaking that's true, but it doesn't mean that you can use `_` and `-` interchangeably.

If you set a key using a `-` it stays a hyphen, even if you later set key with `_`. And likewise, if you first set it using a `_` it says an underscore.

```
# first session

dask.config.set({"foo.a_b": 123})
dask.config.get("foo")  # {'a_b': 123}

dask.config.set({"foo.a-b": 123})
dask.config.get("foo")  # still a_b, keeps original underscore

# new session

dask.config.set({"foo.a-b": 123})
dask.config.get("foo")  # now it's {'a-b': 123}

dask.config.set({"foo.a_b": 123})
dask.config.get("foo")  # still a-b, keeps original hyphen
```

Maybe this is a bug in `xgboost`, but the code there does `dask.config.get("xgboost", default=None)` and then throws an error if you have `scheduler-address` instead of `scheduler_address`:

https://github.com/dmlc/xgboost/blob/606be9e6639dc3b24a5af1e8a131673f90aabdca/python-package/xgboost/dask.py

Even if this is a bug in `xgboost` it seems like the expectation around `-` and `_` needs to be better communicated and/or enforced.

**Comment by ntabris** [comment]
The docs also say that

```
export DASK_DISTRIBUTED__SCHEDULER__WORK_STEALING=True
export DASK_DISTRIBUTED__SCHEDULER__ALLOWED_FAILURES=5
export DASK_DISTRIBUTED__DASHBOARD__LINK="/user/<user>/proxy/8787/status"
```

results in the same thing as

```
{
    'distributed': {
        'scheduler': {
            'work-stealing': True,
            'allowed-failures': 5
        }
    }
}
```

This is false.

For example, run `DASK_DISTRIBUTED__SCHEDULER__WORK_STEALING=True ipython` and then

```python
import dask

dask.config.config['distributed']  # result: {'scheduler': {'work_stealing': True}}
```

Note the underscore.

**Comment by jsignell** [comment]
I agree that it would be better if there were one canonical version of these config fields, but I do think that the current behavior is intentional. It is described in the `canonical_name` function:

https://github.com/dask/dask/blob/c5e4b7c367a7283c81c49978020a26d04c34d3b5/dask/config.py#L62-L66

In practice in dask it never matters if you have used an underscore or a hyphen because the `get` handles any needed coercion. So in the example that  you provided above:

```python
dask.config.get("foo.a_b")  # 123
dask.config.get("foo.a-b")  #123
```

I think that is the recommended way for accessing nested config values. 

So I believe that limiting the input to only accept hyphens is an xgboost bug.
